packages/strapi-admin/services/permission/sections-builder/handlers.js

```diff
@@ -0,0 +1,137 @@
+'use strict';
+
+const { curry, matchesProperty, pick } = require('lodash/fp');
+
+const utils = {
+  isOfKind: kind => matchesProperty('kind', kind),
+  resolveContentType: uid => strapi.contentTypes[uid],
+  isNotInSubjects: subjects => uid => !subjects.find(subject => subject.uid === uid),
+  hasProperty: curry((property, subject) => {
+    return !!subject.properties.find(prop => prop.value === property);
+  }),
+  getValidOptions: pick(['applyToProperties']),
+};
+
+/**
+ * Transforms & adds the given action to the section's actions field
+ * Note: The action is transformed to a setting specific format
+ */
+const settings = (action, section) => {
+  const { category, subCategory, displayName, actionId, options } = action;
+
+  section.actions.push({
+    label: displayName,
+    category,
+    subCategory,
+    actionId,
+    ...utils.getValidOptions(options),
+  });
+};
+
+/**
+ * Transforms & adds the given action to the section's actions field
+ * Note: The action is transformed to a plugin specific format
+ */
+const plugins = (action, section) => {
+  const { plugin, subCategory, displayName, actionId, options } = action;
+
+  section.actions.push({
+    label: displayName,
+    plugin,
+    subCategory,
+    actionId,
+    ...utils.getValidOptions(options),
+  });
+};
+
+/**
+ * Transforms & adds the given action to the section's actions field
+ * Note: The action is transformed to a content-type specific format
+ */
+const contentTypesBase = (action, section) => {
+  const { displayName, actionId, subjects, options } = action;
+
+  section.actions.push({
+    label: displayName,
+    actionId,
+    subjects,
+    ...utils.getValidOptions(options),
+  });
+};
+
+/**
+ * Initialize the subjects array of a section based on the action's subjects
+ */
+const subjectsHandlerFor = kind => (action, section) => {
+  const { subjects } = action;
+
+  const toSubjectTemplate = ct => ({ uid: ct.uid, label: ct.info.name, properties: [] });
+
+  const newSubjects = subjects
+    // Ignore already added subjects
+    .filter(utils.isNotInSubjects(section.subjects))
+    // Transform UIDs into content-types
+    .map(utils.resolveContentType)
+    // Only keep specific kind of content-types
+    .filter(utils.isOfKind(kind))
+    // Transform the content-types into section's subjects
+    .map(toSubjectTemplate);
+
+  section.subjects.push(...newSubjects);
+};
+
+const buildDeepAttributesCollection = attributes => {
```

Comment:
Should we keep track of some kind of depth level here (circular dependencies: ctA => compo => ctA => compo => ...)?